veps/sig-compute/hypervisor-abstraction.md

```diff
@@ -0,0 +1,590 @@
+          // Adjust memlock limit for QEMU process
+          AdjustResources(vmi *v1.VirtualMachineInstance, additionalOverheadRatio *string) error
+          // Assign housekeeping CPU threads to housekeeping cgroup
+          HandleHousekeeping(vmi *v1.VirtualMachineInstance, domain *api.Domain) error
+      }
+      ```
+
+  2. `HypervisorCapabilityExtractor` interface for querying the hypervisor device for its capabilities.
+      ```go
+      type HypervisorCapabilityExtractor interface {
+        GetHypervFeatures() []string
+      }
+      ```
+
+  3. `VirtLauncherResourceRenderer` for computing resource requirement of the `virt-launcher` pod.
+      ```go
+      type VirtLauncherResourceRenderer interface {
+        GetMemoryOverhead(vmi *v1.VirtualMachineInstance, arch string, additionalOverheadRatio *string) resource.Quantity
+      }
+      ```
+          
+
+- **Converter library (`pkg/virt-launcher/virtwrap/converter/`)** – Adds a new `Converter` interface, which contains the main function to convert VMI to Libvirt domain.
+
+  ```golang
+  type Converter interface {
+    Convert_v1_VirtualMachineInstance_To_api_Domain(vmi *v1.VirtualMachineInstance, domain *api.Domain, c *ConverterContext) (err error)
+  }
+  ```
+
+  The `BaseConverter` struct that implements the above interface would contain common functionality for VMI to domain conversion. Each hypervisor will be able to implement their own converter, e.g., `MshvConverter`, embed the `BaseConverter` and override certain functions of the `BaseConverter` with their own.
+
+- **Node labeller** – Adds parameters `--virt-type` and `--hypervisor-device` to the `cmd/virt-launcher/node-labeller/node-labeller.sh` script, so that it can correctly query Libvirt for node and domain capabilities.
+
+This split preserves the “implement once, reuse everywhere” story without routing everything through a monolithic interface. New hypervisors can land incrementally—start with defaults and webhooks, add converter support, then extend node labelling—while keeping the contract for each area explicit and testable.
+
+### Selection
+
+- `virt-config` loads cluster-wide defaults from an additive `hypervisor` field on the `KubeVirt` CR. The `hypervisor` field is a list of hypervisors that can be supported on the cluster. 
+
+  **Single-hypervisor Constraint:** In the current VEP, we will enforce that the number of elements in this list is less than or equal to 1, i.e., to enforce only a single hypervisor for the entire cluster. A future VEP will consider adding support for multiple hypervisors on the same cluster. The `name` in each hypervisor configuration entry selects the cluster-wide hypervisor implementation. Supporting multiple hypervisors in the same cluster will also necessitate the addition of a per-VMI field `hypervisor` to denote on which hypervisor the VMI has to be created.
+  
+  A dedicated feature gate, `ConfigurableHypervisor`, guards the new functionality:
```

Comment:
The VEP suggests to perform a refactor of the current code, to extract existing KVM assumption into a modular code, allowing additional hypervisors to be used instead. This refactoring does not require a FG as there is no functionality to protect.
New hypervisors are possible only if one deploys or configures something explicitly. That, and only that, will trigger a change in underlying behavior to use a different hypervisor.
Therefore, the question that comes to mind is: What will the FG protect against? It will be another knob to just ignore other hypervisors? Can we make this explicit?